ShayanMahnam/lentegeur-hospital-facility-board
# 85 - Content Fetching - Baz

## Description

Now the Strapi REST API is enabled and live, eg: https://lhfb.onrender.com/api/services
We should move towards fetching as much of our website content from the backend as possible. 
To allow us to use and showcase the content management aspect of our project which is crucial for handing off our website to the Product Owner.

I have chosen (and tried to preliminarily learn) [TanStack ReactQuery](https://tanstack.com/query/latest) as a good tool to help us handle all the fetching (and state management/cacheing of that data).
The implementation is currently very basic.

## Complete:

- Added placeholder Content to Strapi REST API to allow testing of this PR (Services, News Articles, Events)
- Adjusted HomePage,  ServicesPage, NewsPage, EventsPage, GalleryPage to use React Router to fetch content from the Live Strapi REST API
- Adjusted Sub Pages to fetch content from the Live Strapi REST API
- Created contentParser utility function to convert a Strapi `content` HTML String into structured React components
- Used the contentParser function to correctly render the `content` to the DOM

## Issues

- Need to implement better error handling if the Promise resolves via React Query because of the specific Strapi style of responses for Errors
- Need to look into prefetching (to try to eliminate `loading` states as much as possible(?)) but was advised this is heavy on the client and maybe ill advised
- Need to understand how to reduce or minimise the queries on a main listing page to subpage as data is somewhat similar

## How to Test

- `npm install` to make sure you have the 2 added packages `@tanstack/react-query` and `html-react-parser`
- Click through on Each of the Pages
- Notice that the first time you go to any of the above Pages there is a Loading state
- But if you go somewhere else and come back, there is no longer a Loading State, because the data has been Cached

- For the parsing inspect the structure of the DOM Tree to see the Content has been built correctly as separate `<p>` elements for example

## Related to

Fixes #85

## Checklist:

- [X] I have carefully reviewed my own code
- [X] ~I have updated any documentation~
